### ISSUE TITLE
Division by a non-constant in symmetric cryptography code

### ISSUE THREAD

**Issue by gilles-peskine-arm** [ARMmbed/mbedtls#3654]
Some processors do not have hardware division, and therefore the compiler generates code using other operations (additions, multiplications, shifts) for division by a constant, and calling a toolchain-provided function for division by a variable. In some environments, Mbed TLS is linked into a code segment that doesn't include the division function. Therefore my understanding is that we limit the use of division to asymmetric cryptography (which really needs division), and only for `mbedtls_mpi_uint`, not for larger types.

However, looking through the code base, I see other uses of division by a non-constant.
```
grep -n -P ' / *(?![ */0-9A-Z]|sizeof|[^"]*")' library/*.c
```
shows on 2.24.0 (in addition to many false positives):
```
library/cmac.c:276:    n = ( ilen + block_size - 1 ) / block_size;
library/ecp.c:1886:        i = 1U << ( j / d );
library/ecp.c:2296:    d = ( grp->nbits + w - 1 ) / w;
library/hkdf.c:110:    n = okm_len / hash_len;
library/memory_buffer_alloc.c:240:    if( n == 0 || size == 0 || len / n != size )
library/psa_crypto.c:4271:            / operation->block_size * operation->block_size;
library/ssl_tls.c:1991:        uint32_t ratio = ssl->conf->hs_timeout_max / ssl->conf->hs_timeout_min + 1;
library/timing.c:256:                                 / hfreq.QuadPart );
library/timing.c:499:    ratio = cycles / millisecs;
library/timing.c:509:        if( cycles / millisecs < ratio - ratio / 5 ||
library/timing.c:510:            cycles / millisecs > ratio + ratio / 5 )
```

The types of the dividends are:
* `w` in `ecp.c`: `unsigned char`
* Timeout in `ssl_tls.c`: `uint32_t`.
* `QuadPart` in `timing.c`: doesn't matter, this is Windows code and harware division is available.
* `millisecs` in `timing.c`: `unsigned long`. Only in a self-test.
* Others: `size_t`.

This is probably not a problem in practice since nobody has complained. On most platforms, `mbedtls_mpi_uint` is `size_t` (or perhaps twice as large if `MBEDTLS_NO_UDBL_DIVISION` is defined). Since most of the occurrences above are `size_t` or `uint32_t`, it would only be a problem in a build without bignum (so either no asymmetric cryptography, or only through alt implementations, and I haven't checked whether the divisions in `ecp.c` can be avoided with alt implementations).

Should we fix this? Or should we stop worrying about division on a type that's bounded by at least one of `uint32_t`, `size_t`, `unsigned long` or `mbedtls_mpi_uint`?

Regardless of our decision, we should document the expectations precisely. The only documentation I'm aware of is that of [`MBEDTLS_NO_UDBL_DIVISION`](https://github.com/ARMmbed/mbedtls/blob/mbedtls-2.24.0/include/mbedtls/config.h#L62), which doesn't say whether division by `size_t` is supposed to be acceptable, or whether the use of division is acceptable when the bignum module is disabled.

**Comment by mpg** [comment]
Note: I think `%` should count as a division as well.

> Or should we stop worrying about division

To be honest, I don't think we've ever consistently worried about division. I know of only one place if the code where we did (the old Lucky 13 counter-measure) and I suspect it was an accident.

I'm not entirely sure about labeling this as a bug, though, IMO it could be considered an enhancement: we currently don't offer an option for division-free builds, we never said we did, and offering such an option would be a new feature. Which of course we should consider only if people are asking for it, and apparently nobody is so far. Considering people already asked for `MBEDTLS_NO_UDBL_DIVISION` and `MBEDTLS_NO_64BIT_MULTIPLICATION`, it looks like at least some people know they can ask if needed.

Or do you think it's a documentation bug that it's undocumented? We say we're using standard C, we're not saying we don't use division, and standard C includes division, so while I agree it's not explicitly documented, I'd still say it shouldn't come as a big surprise. Though I agree that it would be good to have better documentation about which features of C we use and which we stay away from (for example, we'll never use VLAs).

> asymmetric cryptography (which really needs division)

On a less important note, I'd like to nuance that assertion: with appropriate design choices, I'm positive you can do all of ECC without division. I'm almost sure you can to DHM key exchange and RSA public/private operations without it too. You might need it for DHM parameter generation and RSA key generation, though, I'm not sure about those parts.

**Comment by gilles-peskine-arm** [comment]
> we currently don't offer an option for division-free builds

Right, all we offer is a build without “double-width integer division”. Which we don't quite respect since some of the divisions are “double-width” on some platforms. Since we only support 32-bit and 64-bit platforms (we do not support 16-bit platforms and any other size is too exotic for us to have an official position), this means that according to the documentation, with ``MBEDTLS_NO_UDBL_DIVISION` enabled, in practice, division on `size_t` or `uint32_t` is ok but division by `unsigned long` depends on whether you consider IP32L64 to be too exotic for concern or not.

> We say we're using standard C, we're not saying we don't use division

We don't actually say that we don't use floating point, either, but I'd be annoyed if linking my application to a crypto/tls library on a platform without hardware floating point pulled in a floating point library. Integer division is less clear-cut, but it is a practical problem that I've faced before (not in Mbed TLS) where a module that used symmetric crypto and contained code that divided by the block size had a minor refactoring where the compiler suddenly stopped realizing that the block size was a constant, and this caused the linking to fail because the linker was configured with an explicit list of toolchain functions to include. So if we were not using integer division and started to do it, this could break existing applications.

> with appropriate design choices, I'm positive you can do all of ECC without division

Well, yes, you can .
